### PR TITLE
Another threadpool: Avoid creating hundreds of threads in GGML

### PR DESCRIPTION
**Pull request by besnardjb**
# TL/DR:

When profiling the code I noticed hundreds of threads were created, making profiling not practical and maybe leading to some performance overhead. I then attempted to make a simple threadpool inside GGML to see if it made any difference at all. Gains at small scale are not notable, and inference does not change much. However, the more threads the more overhead and more importantly, the code is easier to profile (not 100s of threads to look at).

This is my first attempt at hacking the code, there might be cleaner attempts or better ways, I have seen PR #710, #851 which are similar. I'm also not sure of how to properly benchmark this, did launch `./bin/benchmark`.

## Perf on x86-64 (Genoa)

- One socket:
   - ./bin/benchmark -t 96 -i 100
      - without: (last line) average flops 1497.53
      - with : (last line) average flops 2970.68
- Two sockets:
   - ./bin/benchmark -t 192 -i 100
      - without: (last line) average flops 1044.79
      - with : (last line) average flops 3488.96

## Perf on x86-64 (Ivy Bridge)

- One socket:
   - ./bin/benchmark -t 18 -i 100
      - without: (last line) average flops 367.60
      - with : (last line) average flops 378.21
- Two sockets:
   - ./bin/benchmark -t 36 -i 100
      - without: (last line) average flops 633.22
      - with : (last line) average flops 706.8

### CODE REVIEW

**Comment by slaren** [comment]
You can use `llama-bench` to test the overall performance. I ran a test with and without KV offload since this is one case where the threads are started multiple times in each evaluation.

`LLAMA_CUDA=1 scripts/compare-commits.sh master thread_pool -nkvo 0,1`

| GPU         | Model         |   NKVO | Test        |   t/s master |   t/s thread_pool |   Speedup |
|:------------|:--------------|-------:|:------------|-------------:|------------------:|----------:|
| RTX 3090 Ti | llama 7B Q4_0 |      0 | pp512       |      4313.72 |           4525.64 |      1.05 |
| RTX 3090 Ti | llama 7B Q4_0 |      0 | tg128       |       154.96 |            154.17 |      0.99 |
| RTX 3090 Ti | llama 7B Q4_0 |      0 | pp512+tg128 |       622.75 |            625.94 |      1.01 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | pp512       |       619.01 |            527.14 |      0.85 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | tg128       |        21.55 |             49.80 |      2.31 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | pp512+tg128 |        84.56 |            135.97 |      1.61 |

**Comment by besnardjb** [comment]
Thanks for the pointer to `llama-bench`! sorry for missing it. I did run a few CPU tests on a bi-socket machine. I was not able to use the convenient `compare-commits.sh` script yet, as there seems to be some warning output piped in SQL.

Overall, there seem to be punctual gains, but it's close to measurement noise. I've also cleaned up the code a bit and tried to fix the Windows build ( :warning: without access to a Windows machine, unfortunately). For Windows, as apparently there is no direct equivalent of a semaphore, I have used busy waiting for work in threads to skip dependencies (measured its impact below on Linux too, just to see what it does).

Ran: `./bin/llama-bench -m ../models/7B/mistral-7b-instruct-v0.2.Q4_0.gguf -nkvo 0,1`

## ThreadPool (Semaphore)

```
| model                          |       size |     params | backend    |    threads |          nkvo |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         pp512 |    248.15 ± 1.38 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         tg128 |     22.39 ± 0.20 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |   pp512+tg128 |     81.91 ± 1.01 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         pp512 |    211.54 ± 8.78 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         tg128 |     19.73 ± 0.17 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |   pp512+tg128 |     70.52 ± 0.57 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         pp512 |   241.72 ± 10.81 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         tg128 |     22.25 ± 0.45 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |   pp512+tg128 |     82.73 ± 0.39 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         pp512 |    227.08 ± 2.27 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         tg128 |     19.42 ± 0.14 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |   pp512+tg128 |     70.71 ± 0.77 |
```


## Threadpool (ACTIVE Poll -- as for windows measured on Linux)

```
| model                          |       size |     params | backend    |    threads |          nkvo |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         pp512 |    246.28 ± 0.34 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         tg128 |     24.11 ± 0.10 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |   pp512+tg128 |     84.40 ± 0.54 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         pp512 |    224.57 ± 0.51 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         tg128 |     19.38 ± 0.23 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |   pp512+tg128 |     71.38 ± 0.79 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         pp512 |    247.67 ± 1.07 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         tg128 |     24.73 ± 0.48 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |   pp512+tg128 |     83.10 ± 0.75 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         pp512 |    236.53 ± 1.82 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         tg128 |     19.02 ± 0.08 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |   pp512+tg128 |     69.73 ± 0.14 |
```


# Master (9ca4abed893685692f90413e4d43153af12342d9)

```
| model                          |       size |     params | backend    |    threads |          nkvo |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ------------: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         pp512 |   246.89 ± 11.27 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |         tg128 |     19.84 ± 0.01 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             0 |   pp512+tg128 |     75.51 ± 0.43 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         pp512 |    238.93 ± 6.04 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |         tg128 |     18.17 ± 0.15 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             0 |   pp512+tg128 |     69.23 ± 0.48 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         pp512 |   253.94 ± 14.74 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |         tg128 |     19.85 ± 0.02 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |         96 |             1 |   pp512+tg128 |     75.34 ± 1.20 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         pp512 |   236.17 ± 16.98 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |         tg128 |     18.25 ± 0.08 |
| llama 7B Q4_0                  |   3.83 GiB |     7.24 B | CPU        |        192 |             1 |   pp512+tg128 |     69.36 ± 0.58 |
```

**Comment by slaren** [comment]
To use `compare-commits.sh` you need to have `sqlite3` installed. The nkvo parameter only has an effect when offloading a model to a GPU, basically what happens with nkvo is that the attention part of each layer of run on the CPU and the rest on the GPU, and this causes a call to `ggml_graph_compute` on every layer, so the threads have to be launched once per layer.

**Comment by besnardjb** [comment]
Got it, I basically an the same test twice :laughing: ... For `compare-commits.sh` I did install sqlite and gitpython but I had two issues (pure CPU run again):

- In ggml.c I had a warning (namely: `GGML_PRINT("WARNING: /proc/sys/kernel/numa_balancing is enabled, this has been observed to impair performance\n");`) it was sent to stdout and ended up mixed with the SQL; This may go to stderr and the script may only capture stdout.

- Then tried commenting the line out to see, but ended up with out of range error in `scripts/compare-llama-bench.py` on lines 302-303 and did not take the time to investigate more.

```
        gpu_blas = bool(rows_full[0][KEY_PROPERTIES.index("gpu_blas")])
        ngl = int(rows_full[0][KEY_PROPERTIES.index("n_gpu_layers")])
``` 

Thank you will further look into it.

**Comment by besnardjb** [comment]
It is a very good idea. Will look into it after finding myself a windows VM. Also I'm currently not freeing threads will need to refcount init and release.

**Comment by besnardjb** [comment]
After looking into the Futex approach I realized it was about to add a lot of ifdefs and several lines of (relatively complex) codes. I then thought that I was in fact trying to implement an OpenMP parallel region. And thus, resorted to trying OpenMP directly. This basically translated in a single `#pragma omp` and avoids manual handling. In addition, what is nice is that it could be used on other for loops too elsewhere in the code without repeating the whole threadpool thingy (the runtime does it for free). I think it ends up simpler, but all comments are welcome of course.

I hope it works well on windows, lets see. 

Also I realized performance counters in `ggml_graph_compute` were not accounting for thread creation cost, I thus moved them a few line up on master to do the same measure in the two case.


## 72 Cores  (moved timings for thread creation)

| CPU                                       | Model         | Test        |   t/s master |   t/s openmp |   Speedup |
|:------------------------------------------|:--------------|:------------|-------------:|-------------:|----------:|
| Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz | llama 7B Q4_0 | pp512       |        47.17 |        58.55 |      1.24 |
| Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz | llama 7B Q4_0 | tg128       |         9.08 |        13.96 |      1.54 |
| Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz | llama 7B Q4_0 | pp512+tg128 |        27.37 |        35.15 |      1.28 |

## 192 cores (moved timings for thread creation)

| CPU                             | Model         | Test        |   t/s master |   t/s openmp |   Speedup |
|:--------------------------------|:--------------|:------------|-------------:|-------------:|----------:|
| AMD EPYC 9654 96-Core Processor | llama 7B Q4_0 | pp512       |       226.86 |       241.64 |      1.07 |
| AMD EPYC 9654 96-Core Processor | llama 7B Q4_0 | tg128       |        18.24 |        21.42 |      1.17 |
| AMD EPYC 9654 96-Core Processor | llama 7B Q4_0 | pp512+tg128 |        69.03 |        76.94 |      1.11 |

# Old Laptop 4 cores (moved timings for thread creation)
| CPU                                       | Model          | Test        |   t/s master |   t/s openmp |   Speedup |
|:------------------------------------------|:---------------|:------------|-------------:|-------------:|----------:|
| Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz | phi3 3B Q4_K_M | pp512       |        13.07 |        13.28 |      1.02 |
| Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz | phi3 3B Q4_K_M | tg128       |         7.51 |         7.60 |      1.01 |
| Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz | phi3 3B Q4_K_M | pp512+tg128 |        11.38 |        11.28 |      0.99 |


# H100 (moved timings for thread creation)

| GPU              | Model         |   NKVO | Test        |   t/s master |   t/s openmp |   Speedup |
|:-----------------|:--------------|-------:|:------------|-------------:|-------------:|----------:|
| NVIDIA H100 PCIe | llama 7B Q4_0 |      0 | pp512       |      6366.18 |      6222.38 |      0.98 |
| NVIDIA H100 PCIe | llama 7B Q4_0 |      0 | tg128       |       130.68 |       120.55 |      0.92 |
| NVIDIA H100 PCIe | llama 7B Q4_0 |      0 | pp512+tg128 |       549.15 |       515.65 |      0.94 |
| NVIDIA H100 PCIe | llama 7B Q4_0 |      1 | pp512       |       601.44 |       711.63 |      1.18 |
| NVIDIA H100 PCIe | llama 7B Q4_0 |      1 | tg128       |         4.94 |        33.61 |      6.81 |
| NVIDIA H100 PCIe | llama 7B Q4_0 |      1 | pp512+tg128 |        22.21 |       128.38 |      5.78 |

File: ggml.c [hunk 1]
```diff
@@ -20086,44 +20097,50 @@ enum ggml_status ggml_graph_compute(struct ggml_cgraph * cgraph, struct ggml_cpl
...
-                .ith = j,
-                .shared = &state_shared,
-                .ec = GGML_STATUS_SUCCESS,
-            };
+    const int64_t perf_start_cycles  = ggml_perf_cycles();
+    const int64_t perf_start_time_us = ggml_perf_time_us();
 
+    /* Loop is reversed as in the NO_OMP case we want threads to start
+    before the main thread (j==0) */
+    #pragma omp parallel for shared(workers,state_shared)
```
Comment (slaren) [inline ggml.c:20105]: Might need to add a `num_threads(n_threads)` here to make sure that omp always launches all the threads, otherwise it will deadlock.

**Comment by slaren** [comment]
Unfortunately the omp is slower for me, and it causes a significant performance regression without nkvo.

| GPU         | Model         |   NKVO | Test        |   t/s master |   t/s thread_pool |   Speedup |
|:------------|:--------------|-------:|:------------|-------------:|------------------:|----------:|
| RTX 3090 Ti | llama 7B Q4_0 |      0 | pp512       |      4512.80 |           4076.89 |      0.90 |
| RTX 3090 Ti | llama 7B Q4_0 |      0 | tg128       |       154.23 |            140.59 |      0.91 |
| RTX 3090 Ti | llama 7B Q4_0 |      0 | pp512+tg128 |       627.11 |            570.19 |      0.91 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | pp512       |       526.36 |            489.66 |      0.93 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | tg128       |        21.86 |             42.83 |      1.96 |
| RTX 3090 Ti | llama 7B Q4_0 |      1 | pp512+tg128 |        84.62 |            131.32 |      1.55 |

File: ggml.c (hunk 1 above)
Comment (besnardjb) [reply ggml.c:20105]: You are right it is indeed fragile. I'm not event sure setting thread count covers all cases (https://www.openmp.org/spec-html/5.0/openmpsu35.html#x55-880002.6.1).

**Comment by besnardjb** [comment]
Thank you very much for checking. Indeed turns out it cannot be so simple; it is hard to beat those hundreds of threads ! 

As Pools did perform better on GPU configs, *maybe* OMP is doing some busy waiting in between parallels interfering with Cuda. I will try first to fix the parallel for risky logic and will play a bit more with it out of curiosity.

**Comment by besnardjb** [comment]
Superseded by #7606 using Openmp :+1:. I still think that there are gains to have at the level of the barriers `ggml_graph_compute_thread_sync_node` & `ggml_graph_compute_thread_sync_task`.
